lib/style-engine/class-wp-style-engine-gutenberg.php

```diff
@@ -0,0 +1,257 @@
+				'value_func'   => 'gutenberg_get_style_engine_css_box_rules',
+			),
+			'margin'  => array(
+				'property_key' => 'margin',
+				'value_func'   => 'gutenberg_get_style_engine_css_box_rules',
+			),
+		),
+	);
+
+	/**
+	 * Register action for outputting styles when the class is constructed.
+	 */
+	public function __construct() {
+		// Borrows the logic from `gutenberg_enqueue_block_support_styles`.
+		$action_hook_name = 'wp_footer';
+		if ( wp_is_block_theme() ) {
+			$action_hook_name = 'wp_enqueue_scripts';
+		}
+		add_action(
+			$action_hook_name,
+			array( $this, 'output_styles' )
+		);
+	}
+
+	/**
+	 * Utility method to retrieve the main instance of the class.
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @return WP_Style_Engine_Gutenberg The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Resets the style store.
+	 */
+	public function reset() {
+		$this->registered_styles = array();
+	}
+
+	/**
+	 * Compiles and returns styles to be included in an HTML document.
+	 *
+	 * @return string CSS rules to embed in a `<style />`.
+	 */
+	public function get_generated_styles() {
+		$output = '';
+		foreach ( $this->registered_styles as $selector => $rules ) {
+			$output .= "{$selector} {\n";
+
+			if ( is_string( $rules ) ) {
+				$output .= '  ';
+				$output .= $rules;
+			} else {
+				foreach ( $rules as $rule => $value ) {
+					$output .= "  {$rule}: {$value};\n";
+				}
+			}
+			$output .= "}\n";
+		}
+		return $output;
+	}
+
+	/**
+	 * Stores style rules for a given CSS selector (the key) and returns an associated classname.
+	 * Accepts an array of options, rules, and selector for constructing the rules.
+	 *
+	 * @param string $key     A class name used to construct a key.
+	 * @param array  $options = array(
+	 *        'prefix'    => (string) An optional beginning classname fragment, such as selector syntax, e.g., `#`
+	 *        'suffix'    => (string) An optional ending classname fragment.
+	 *        'selector'  => (string) An optional CSS selector for specificity, e.g., `> p`.
+	 *        'rules'     => (array) An array of valid CSS rules, e.g., array( 'padding-top' => '1em', 'color' => 'pink' ).
+	 *        'obfuscate' => (boolean) Whether to hash the classname.
+	 *     );.
+	 *
+	 * @return string The class name for the added style.
+	 */
+	public function add_style( $key, $options ) {
+		$class    = ! empty( $options['suffix'] ) ? $key . '-' . sanitize_title( $options['suffix'] ) : $key;
+		$selector = ! empty( $options['selector'] ) ? ' ' . trim( $options['selector'] ) : '';
+		$rules    = ! empty( $options['rules'] ) ? $options['rules'] : array();
+		$prefix   = ! empty( $options['prefix'] ) ? $options['prefix'] : '.';
+
+		if ( ! $class ) {
+			return;
+		}
+
+		if ( ! empty( $options['obfuscate'] ) && true === $options['obfuscate'] ) {
```

Comment:
I know you're still working on this, but just wanted to say I love this idea for where to inject the obfuscation — it neatly combines our two approaches by using the unique / semantic classname as the hash, so it's an easy switch between the two approaches ✨

Reply:
Thanks! Ultimately this was a side-experiment inside a side-experiment 😆 

It'd be nice to follow up on thought. 

Still, the minimum I suppose we could get from this PR is:

- replicating the current style engine functionality (padding and inline styles)
- tests (when I could be bothered writing them)

Not sure yet about moving the PHP file to the package. It's pretty tightly coupled to Gutenberg and there's no precedent for it yet. 

What do you reckon?

Reply:
> What do you reckon?

Conceptually, I think I liked the idea of keeping it in the package to begin with, but it's since become clearer that it'll get called from lots of other places and will be more tightly integrated into other Gutenberg files. I think I'd keep it in `lib/` for the time-being.